samples/bluetooth/bap_broadcast_sink/src/audio_codec.c

```diff
@@ -0,0 +1,172 @@
+
+/*
+ * Total size of the ring buffer used to queue audio data for the codec.
+ *
+ * The buffer is sized to hold 20 CODEC_BLOCK_SIZE blocks. Having multiple
+ * blocks queued reduces the likelihood of underruns when the producer is
+ * briefly delayed, while still keeping overall latency and RAM usage
+ * bounded. Increasing the multiplier adds buffering (and latency / RAM
+ * consumption); decreasing it reduces buffering but makes underruns more
+ * likely on a busy system.
+ */
+#define RING_BUF_SIZE    (CODEC_BLOCK_SIZE * 20)
+
+#if DT_NODE_HAS_STATUS_OKAY(DT_ALIAS(codec0))
+static uint8_t ring_buffer[RING_BUF_SIZE];
+static struct ring_buf rb;
+static bool codec_configed;
+static uint8_t block_data[CODEC_BLOCK_SIZE];
+static const struct device *g_dev;
+
+static void tx_done(const struct device *dev, void *user_data)
+{
+	uint32_t avail;
+	uint32_t read;
+
+	avail = ring_buf_size_get(&rb);
+	if (avail < CODEC_BLOCK_SIZE) {
+		LOG_WRN("Buffer empty\n");
+		return;
+	}
+	read = ring_buf_get(&rb, block_data, CODEC_BLOCK_SIZE);
+	if (read != CODEC_BLOCK_SIZE) {
+		LOG_WRN("Failed to read full block from ring buffer: %u bytes\n", read);
+		return;
+	}
+	audio_codec_write(dev, block_data, CODEC_BLOCK_SIZE);
+}
+int audio_codec_open(void)
+{
+	ring_buf_init(&rb, RING_BUF_SIZE, ring_buffer);
+	g_dev = DEVICE_DT_GET(DT_ALIAS(codec0));
+	if (!device_is_ready(g_dev)) {
+		LOG_ERR("Could not open codec device\n");
+		return -EIO;
+	}
+	return 0;
+}
+int audio_codec_cfg(uint32_t samplerate)
+{
+	int ret;
+	audio_property_value_t val;
+	struct audio_codec_cfg cfg = {
+		.dai_type = AUDIO_DAI_TYPE_PCM,
+		.dai_cfg.pcm.dir = AUDIO_DAI_DIR_TX,
+		.dai_cfg.pcm.pcm_width = AUDIO_PCM_WIDTH_16_BITS,
+		.dai_cfg.pcm.channels = 1,
+		.dai_cfg.pcm.block_size = CODEC_BLOCK_SIZE,
+		.dai_cfg.pcm.samplerate = samplerate,
+	};
+
+	if (codec_configed) {
+		LOG_WRN("Already configured\n");
+		return -EALREADY;
+	}
+	LOG_INF("codec: samplerate=%d block_size=%d\n", samplerate, CODEC_BLOCK_SIZE);
+	audio_codec_register_done_callback(g_dev, tx_done, NULL, NULL, NULL);
+	ret = audio_codec_configure(g_dev, &cfg);
+	if (ret != 0) {
+		LOG_ERR("Failed to configure codec (err %d)\n", ret);
+		return ret;
+	}
+	ret = audio_codec_start(g_dev, AUDIO_DAI_DIR_TX);
+	if (ret != 0) {
+		LOG_ERR("Failed to start codec (err %d)\n", ret);
+		return ret;
+	}
+	val.vol = SPEAKER_VOL;
+	ret = audio_codec_set_property(g_dev, AUDIO_PROPERTY_OUTPUT_VOLUME, 0, val);
+	if (ret != 0) {
+		LOG_ERR("Failed to set codec output volume (err %d)\n", ret);
+		audio_codec_stop(g_dev, AUDIO_DAI_DIR_TX);
+		return ret;
+	}
+	codec_configed = true;
+
+	return 0;
+}
+uint32_t audio_codec_write_data(const uint8_t *data, uint32_t len)
+{
+	uint32_t bytes_put;
+
+	bytes_put = ring_buf_put(&rb, data, len);
+	if (bytes_put != len) {
+		LOG_WRN("Buffer full\n");
+	}
+	return bytes_put;
+}
+int audio_codec_close(void)
+{
+	if (g_dev && codec_configed) {
+		audio_codec_stop(g_dev, AUDIO_DAI_DIR_TX);
```

Comment:
Not sure why, but I also cannot find the header for `audio_codec_stop`. Is it a `void` function, or can it return an error?